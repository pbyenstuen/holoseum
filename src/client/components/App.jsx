import React, { useEffect } from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import AOS from "aos";
import "aos/dist/aos.css"
import PublicRoute from "../hocs/PublicRoute";
import ProtectedRoute from "../hocs/ProtectedRoute";
import useLoader from "../hooks/useLoader";
import Header from "./Header";
import HomePage from "./HomePage";
import ContactPage from "./ContactPage";
import Footer from "./Footer";
import HologramView from "./HologramView";
import AdminPage from "./AdminPage";
import NotFound from "./NotFound";

const App = ({ api }) => {
    const { data: user, loading, reload: updateUser } = useLoader(async () => await api.auth.getUser());

    useEffect(() => {
        AOS.init({
            once: true,
            duration: 1500
        });
    }, []);

    const LandingLayout = ({ children }) => {
        return (
            <>
                <Header
                    api={api}
                    user={user}
                    updateUser={updateUser} />
                {children}
                <Footer />
            </>
        )
    }

    return (
        <Router>
            <Switch>
                <PublicRoute
                    path={"/hologrammer/:name"}
                    component={HologramView}
                    api={api} />
                <PublicRoute
                    exact path={"/"}
                    component={HomePage}
                    layout={LandingLayout} />
                <PublicRoute
                    path={"/kontakt"}
                    component={ContactPage}
                    layout={LandingLayout} />
                <ProtectedRoute
                    path={"/admin"}
                    component={AdminPage}
                    layout={LandingLayout}
                    user={user}
                    loading={loading}
                    updateUser={updateUser}
                    api={api} />
                <Route
                    path={"*"}
                    component={NotFound} />
            </Switch >
        </Router >
    )
}

export default App;