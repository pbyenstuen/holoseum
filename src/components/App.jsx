import React from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import HeaderBar from "./HeaderBar";
import HomePage from "./HomePage";
import ContactPage from "./ContactPage";

const App = () => {

    return (
        <Router>
            <>
                <HeaderBar />
                <Switch>
                    <Route exact path="/">
                        <HomePage />
                    </Route>
                    <Route path="/kontakt">
                        <ContactPage />
                    </Route>
                </Switch>
            </>
        </Router>
    )
}

export default App;