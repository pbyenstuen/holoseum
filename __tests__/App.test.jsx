import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import App from "../src/client/components/App";

require('react-router-dom').BrowserRouter = ({ children }) => <div>{children}</div>

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component, entries) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter initialEntries={[entries]}>{component}</MemoryRouter>, container);
    });
}

describe("App", () => {
    it("routes to NotFound component on invalid path", async () => {
        const api = {
            auth: {
                getUser: () => ({
                    username: "tidvis",
                })
            }
        }

        await mountComponent(<App api={api} />, "/nothing");

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h1").textContent).toEqual("404");
    });

    it("can route to correct path", async () => {
        const api = {
            auth: {
                getUser: () => ({
                    username: "tidvis",
                })
            }
        }

        await mountComponent(<App api={api} />, "/");

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h1").textContent).toEqual("HOLOSEUM");
    });

    it("renders login component on attempt to access protected route while unauthenticated", async () => {
        const api = {
            auth: {
                getUser: () => {
                    return undefined;
                }
            }
        }

        await mountComponent(<App api={api} />, "/admin");

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h2").textContent).toEqual("Admin - Logg Inn");
    });

    it("gives access to protected route while authenticated", async () => {
        const api = {
            auth: {
                getUser: () => {
                    return "tidvis";
                }
            }
        }

        await mountComponent(<App api={api} />, "/admin");

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h2").textContent).toEqual("Last opp hologram");
    });
});