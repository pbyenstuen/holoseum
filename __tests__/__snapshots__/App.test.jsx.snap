// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App can route to correct path 1`] = `"<div><header><nav><ul><li><a aria-current=\\"page\\" class=\\"is-active\\" href=\\"/\\">HVA ER?</a></li><li><a href=\\"/kontakt\\">KONTAKT</a></li><li><a href=\\"/admin\\">ADMIN</a></li></ul></nav><button>LOGG UT</button></header><div id=\\"home-page-container\\" class=\\"card-container\\"><div id=\\"home-page-banner\\" class=\\"card\\"><div id=\\"home-page-banner-fade\\"><div id=\\"home-page-banner-content-wrapper\\"><h1>HOLOSEUM</h1><p>Et mobilt mini-museum som gir deg en smakebit av et ekte museum. Sammen med våre samarbeidspartnere har vi designet 3D-modeller av museumsgjenstander som kan vises hvor som helst!<br><br>Museumsgjenstandene vises som et 3D-hologram via et nettbrett. Velg blant de ulike muséene og få tilsendt det du trenger for å utforske norsk historie!<br><br>Bla ned for å se hvordan det fungerer.</p><a href=\\"/kontakt\\"><button id=\\"contact-btn\\" class=\\"ui button\\">KONTAKT OSS</button></a></div></div></div><div id=\\"div-center\\" class=\\"card aos-init aos-animate\\" data-aos=\\"fade-up\\"></div><div id=\\"div-1\\" class=\\"guide-point card aos-init aos-animate\\" data-aos=\\"fade-up\\" data-aos-offset=\\"400\\"></div><div id=\\"div-2\\" class=\\"guide-point card aos-init aos-animate\\" data-aos=\\"fade-up\\" data-aos-delay=\\"500\\"></div><div id=\\"div-3\\" class=\\"guide-point card aos-init aos-animate\\" data-aos=\\"fade-up\\" data-aos-delay=\\"500\\"></div></div><footer><h4>footer</h4></footer></div>"`;

exports[`App gives access to protected route when authenticated 1`] = `"<div><header><nav><ul><li><a href=\\"/\\">HVA ER?</a></li><li><a href=\\"/kontakt\\">KONTAKT</a></li><li><a aria-current=\\"page\\" class=\\"is-active\\" href=\\"/admin\\">ADMIN</a></li></ul></nav><button>LOGG UT</button></header><div id=\\"admin-page-container\\"><section id=\\"upload-form-container\\"><h2>Last opp hologram</h2><div id=\\"upload-form-card\\" class=\\"card\\"><form><div><div><label>Navn <input name=\\"name\\" type=\\"text\\" value=\\"\\"></label></div><div><label>Hologram <input name=\\"file\\" type=\\"file\\" value=\\"\\"></label></div><button type=\\"submit\\">LAST OPP</button></div></form></div></section><section id=\\"hologram-list-container\\"><h2>Opplastinger</h2><div><p><i>Ingenting å vise</i></p></div></section></div><footer><h4>footer</h4></footer></div>"`;

exports[`App renders login component on attempt to access protected route while unauthenticated 1`] = `"<div><div id=\\"login-page-container\\"><div id=\\"login-card\\"><a href=\\"/\\">← Hovedside</a><h2>Admin - Logg Inn</h2><form><div><label> <input name=\\"username\\" type=\\"text\\" placeholder=\\"Brukernavn\\" value=\\"\\"></label></div><div><label> <input name=\\"password\\" type=\\"password\\" placeholder=\\"Passord\\" value=\\"\\"></label></div><button>LOGG INN</button></form></div></div></div>"`;

exports[`App routes to NotFound component on invalid path 1`] = `"<div><div id=\\"not-found-container\\" class=\\"center-full-screen\\"><h1>404</h1><p>Siden du lette etter finnes ikke</p><a href=\\"/\\">Klikk her for å gå til hovedsiden</a></div></div>"`;
